kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -475,16 +493,51 @@ static SchemaChanges computeSchemaChangesById(StructType currentSchema, StructTy
         // from map to array.
         continue;
       }
-      StructField updatedField = newElement.getField();
-      if (!existingField.equals(updatedField)) {
+      StructField newField = newElement.getField();
+      List<TypeChange> existingTypeChanges = existingField.getTypeChanges();
+      if (!existingTypeChanges.isEmpty() && newField.getTypeChanges().isEmpty()) {
+        // Copy over type changes from existing field because new schemas
+        // might be constructed elsewhere and not have the persisted type
+        // change metadata.
+        newField = newField.withTypeChanges(existingTypeChanges);
+        newElement.updateField(newField);
+        newSchemaHasUpdates = true;
+      }
+
+      // Ensure the schemas are equal now, updating type changes from clients is not supported
+      // so they should be.
+      if (!existingTypeChanges.equals(newField.getTypeChanges())) {
+        throw new KernelException(
+            String.format(
+                "Detected a modified type changes field at '%s' %s != %s",
+                newElement.getNamePath(), existingTypeChanges, newField.getTypeChanges()));
+      }
+
+      if (!existingField.equals(newField)) {
+        if (!isNestedType(existingField)
+            && !isNestedType(newField)
+            && !existingField.getDataType().equivalent(newField.getDataType())) {
+          List<TypeChange> changes = new ArrayList<>(newField.getTypeChanges());
+          changes.add(new TypeChange(existingField.getDataType(), newField.getDataType()));
+          newElement.updateField(newField.withTypeChanges(changes));
+          newSchemaHasUpdates = true;
+        }
```

Comment:
assert in case of else statement? I know the StructField constructor won't allow it, but just to be sure.

Reply:
I added a comment, but we can't assert here because type differences still need to be detected and handled in schema changes.  I think the cleanup here is to no use equivelant or rather, probably we want equivelantLight to avoid recomputing work for nested types.